File: python/examples/governed_sessions.py
"""Two independent sessions carry a contract through computation and release."""

import tn
from tn.governed import GovernedReader, Session

POLICY = """---
version: 1
schema: tn-agents-policy@v1
---
## research.sample
### instruction
Create an aggregate report.
### use_for
Aggregate research.
### do_not_use_for
Individual disclosure.
### consequences
Contract review.
### on_violation_or_error
Refuse release.
"""


def aggregate(
    wire: str,
    reader: GovernedReader,
    output_session: Session,
    trusted_writer: str,
    approved_policy: str,
) -> tn.GovernedObject:
    source = tn.GovernedObject.parse(wire)
    if source.writer != trusted_writer:
        raise ValueError("source writer is outside this application's accepted writers")

    # The application has approved this exact contract for this operation.
    view = reader.governance(source)
    admitted = view.authorize(
        "aggregate",
        lambda contract, operation: (
            operation == "aggregate"
            and contract.governed_by == trusted_writer
            and contract.policy_ref == approved_policy
        ),
    )
    opened = reader.open(admitted, ["observations"])
    counts = opened.groups["observations"]["counts"]
    if not isinstance(counts, list):
        raise ValueError("observations.counts must contain integer counts")
    total = 0
    for value in counts:
        if type(value) is not int:
            raise ValueError("observations.counts must contain integer counts")
        total += value

    # Derivation carries the source contract and records the admitted operation.
    result = output_session.seal(
        opened.derive("report.generated").group("reports", {"total": total})
    )
    assert opened.object.wire == wire
    assert opened.hidden_groups == ["identities"]
    return result


def main() -> None:
    with tn.Session(POLICY, groups=["observations", "identities"]) as producer:
        with tn.Session(POLICY, groups=["reports"]) as analyst:
            assert producer.did != analyst.did
            approved_policy = producer.policy("research.sample").policy_ref
            source = producer.seal(
                producer.draft("research.sample")
                .group("observations", {"counts": [12, 18]})
                .group("identities", {"names": ["Alice", "Bob"]})
            )
            reader = producer.reader(groups=["tn.agents", "observations"])
            result = aggregate(source.wire, reader, analyst, producer.did, approved_policy)
            contract = analyst.governance(result).governance
            assert result.writer == analyst.did
            assert contract.governed_by == producer.did
            assert contract.policy_ref == approved_policy
            lineage = contract.get("source_lineage")
            assert isinstance(lineage, list) and isinstance(lineage[0], dict)
            assert lineage[0]["object_id"] == source.id

            print("Two independent signing identities")
            print("Opened observations; identities stayed sealed")
            print("Aggregate: 30")
            print("Analyst signed the result with the source contract and lineage")


if __name__ == "__main__":
    main()

File: crypto/tn-core/tests/governed_sessions.rs
#![cfg(feature = "fs")]

use serde_json::json;
use tn_core::runtime::Objects;

const POLICY: &str = "---\nversion: 1\nschema: tn-agents-policy@v1\n---\n## research.sample\n### instruction\nCreate an aggregate report.\n### use_for\nAggregate research.\n### do_not_use_for\nIndividual disclosure.\n### consequences\nContract review.\n### on_violation_or_error\nRefuse release.\n";

#[test]
fn independent_contexts_seal_without_shared_identity_or_reader_material() {
    let alice = Objects::ephemeral(POLICY, "alice.md", &["observations", "identities"]).unwrap();
    let bob = Objects::ephemeral(POLICY, "bob.md", &["observations"]).unwrap();
    assert_ne!(alice.did(), bob.did());
    let source = alice
        .seal(
            alice
                .draft("research.sample")
                .unwrap()
                .group("observations", json!({"count": 42}))
                .unwrap()
                .group("identities", json!({"person": "Alice"}))
                .unwrap(),
        )
        .unwrap();
    assert!(bob.reader().unwrap().governance(&source).is_err());
    let reader = alice.reader_for(["tn.agents", "observations"]).unwrap();
    let admitted = reader
        .governance(&source)
        .unwrap()
        .authorize("aggregate", |_, _| Ok(true))
        .unwrap();
    assert!(reader.open(&admitted, ["identities"]).is_err());
    let opened = reader.open(&admitted, ["observations"]).unwrap();
    let result = bob
        .seal(
            opened
                .derive("report.generated")
                .unwrap()
                .group("observations", json!({"total": 42}))
                .unwrap(),
        )
        .unwrap();
    assert_eq!(result.writer(), bob.did());
    assert_eq!(
        bob.reader()
            .unwrap()
            .governance(&result)
            .unwrap()
            .governance()
            .governed_by(),
        alice.did()
    );
    drop(alice);
    assert!(reader.governance(&source).is_ok());
    assert!(bob.seal(bob.draft("research.sample").unwrap()).is_ok());
}

#[test]
fn context_creation_and_reader_scoping_validate_group_assignments() {
    assert!(Objects::ephemeral(POLICY, "agents.md", &["tn.agents"]).is_err());
    assert!(Objects::ephemeral(POLICY, "agents.md", &["default", "default"]).is_err());
    let objects = Objects::ephemeral(POLICY, "agents.md", &["default"]).unwrap();
    assert!(objects.reader_for(["missing"]).is_err());
    assert!(objects.reader_for(["default", "default"]).is_err());
}
